modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/NestedLoopJoinNode.java

```diff
@@ -673,28 +622,12 @@ public FullOuterJoin(
                 }
             }
 
-            if (waitingRight == 0)
-                rightSource().request(waitingRight = IN_BUFFER_SIZE);
-
-            if (waitingLeft == 0 && leftInBuf.isEmpty())
-                leftSource().request(waitingLeft = IN_BUFFER_SIZE);
-
-            if (requested > 0 && waitingLeft == NOT_WAITING && waitingRight == NOT_WAITING && left == null
-                && leftInBuf.isEmpty() && rightNotMatchedIndexes.isEmpty()) {
-                requested = 0;
-                downstream().end();
-            }
+            tryGetMoreOrEnd(() -> rightNotMatchedIndexes.isEmpty());
```

Comment:
Let's simplify to something:
```
if (rightNotMatchedIndexes.isEmpty() && checkJoinFinished())
  return;

tryToRequestInputs();
```
(WIll look similar to merge join and simplier to read)